src/main/scala/it/unibo/authsim/library/dsl/cryptography/algorithm/hash/HashFactory.scala

```diff
@@ -0,0 +1,24 @@
+package it.unibo.authsim.library.dsl.cryptography.algorithm.hash
+/**
+ * Abstract factory for building hash algorithm
+ */
+object HashAbstractFactory:
+  /**
+   * Enumeration who provide the name of the hash algorithms supported
+   */
+  enum HashFunctionAlgorithm:
+    case MD5, SHA1, SHA256, SHA384
+
+  def apply[A>: HashFunction](name: HashFunctionAlgorithm): A=
+  /**
+   * Apply method for the abstract hash factory
+   * @param name        name of the hash algorithm one wants to create
+   * @tparam A          type of the algorithm
+   * @return            an istance of the algorithm chosen
+   */
+    name match {
+      case HashFunctionAlgorithm.MD5 => new HashFunction.MD5
+      case HashFunctionAlgorithm.SHA1=> new HashFunction.SHA1
+      case HashFunctionAlgorithm.SHA256 => new HashFunction.SHA256
+      case HashFunctionAlgorithm.SHA384 => new HashFunction.SHA384
+    }
```

Comment:
I'm not sure about the usefulness and correctness of this factory: the `enum` can be desynchronized with respect to the implemented `HashFunction`s.

This does not mean it is straight wrong, but it could/should be rethought.

Reply:
I agree with you but the only alternative that came to my mind was to match a string (that would represent the user), however, I would rather have enum not synchronized than to ask user to use a correct name for each hash function